src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -688,10 +692,22 @@ Transactor::operator()()
 
     if (didApply)
     {
+        // Check invariants
+        // if `tecINVARIANT_FAILED` not returned, we can proceed to apply the tx
+        terResult = ctx_.checkInvariants(terResult);
+        if (terResult == tecINVARIANT_FAILED)
+        {
+            // if invariants failed, claim a fee still
+            claimFee(fee, terResult, {});
+            //Check invariants *again* to ensure the fee claiming doesn't
+            //violate invariants.
+            terResult = ctx_.checkInvariants(terResult);
```

Comment:
If `checkInvariants` fails the second time, it returns `tefINVARIANT_FAILED`. `tef` results are not supposed to do anything to the ledger, but this falls through to destroy XRP and apply the tx changes to the view. That means a `tef` result would get written to the ledger, which violates a different invariant. You could do:
```
    if (didApply)
    {
        // Check invariants
        // if `tecINVARIANT_FAILED` not returned, we can proceed to apply the tx
        terResult = ctx_.checkInvariants(terResult);
        if (terResult == tecINVARIANT_FAILED)
        {
            // if invariants failed, claim a fee still
            claimFee(fee, terResult, {});
            //Check invariants *again* to ensure the fee claiming doesn't
            //violate invariants.
            terResult = ctx_.checkInvariants(terResult);
            didApply = isTecClaim(terResult);
        }
    }

    if (didApply)
    {
        // Transaction succeeded fully or (retries are
        // not allowed and the transaction could claim a fee)

        if (!view().open())
        {
[...]
```